src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -33,140 +32,192 @@ internal partial class SyntacticClassificationTaggerProvider
         /// will determine which sections of the file changed and we will use that to notify the editor
         /// about what needs to be reclassified.
         /// </summary>
-        internal partial class TagComputer
+        internal partial class TagComputer : ForegroundThreadAffinitizedObject
         {
+            private readonly SyntacticClassificationTaggerProvider _taggerProvider;
             private readonly ITextBuffer _subjectBuffer;
-            private readonly WorkspaceRegistration _workspaceRegistration;
-            private readonly AsynchronousSerialWorkQueue _workQueue;
+            private readonly IAsynchronousOperationListener _listener;
+            private readonly ClassificationTypeMap _typeMap;
 
-            // this will cache previous classification information for a span, so that we can avoid
-            // digging into same tree again and again to find exactly same answer
-            private readonly LastLineCache _lastLineCache;
+            private readonly WorkspaceRegistration _workspaceRegistration;
 
-            // The latest data about the document being classified that we've cached.  objects can 
-            // be accessed from both threads, and must be obtained when this lock is held.
-            private readonly object _gate = new();
-            private ITextSnapshot _lastProcessedSnapshot;
-            private Document _lastProcessedDocument;
+            private readonly CancellationTokenSource _disposalCancellationSource = new();
 
             /// <summary>
-            /// Data stored by the <see cref="IClassificationService"/> that can be computed in the bg which it will
-            /// benefit from when it is called back for classifications later.
+            /// Work queue we use to batch up notifications about changes that will cause
+            /// us to classify.  This ensures that if we hear a flurry of changes, we don't
+            /// kick off an excessive amount of background work to process them.
             /// </summary>
-            private object _lastProcessedCachedData;
-
-            private Workspace _workspace;
-            private readonly CancellationTokenSource _reportChangeCancellationSource = new();
-
-            private readonly IAsynchronousOperationListener _listener;
-            private readonly IForegroundNotificationService _notificationService;
-            private readonly ClassificationTypeMap _typeMap;
-            private readonly SyntacticClassificationTaggerProvider _taggerProvider;
+            private readonly AsyncBatchingWorkQueue<ITextSnapshot> _workQueue;
```

Comment:
moved to an ABWQ.  It's a really suitable concurrency primitive for this use case.  Effectively, it allows us to get a torrent of change notifications which bubble out to just a single processing operation at a regular cadence (i.e. we always get to the notifications in a reasonable time, but we batch them so we don't do excessive amounts of wasted work).